gene/database/postgresql.py

```diff
@@ -0,0 +1,868 @@
+        DELETE FROM gene_associations WHERE id IN (
+            SELECT ga.id FROM gene_associations ga LEFT JOIN gene_concepts gc
+                ON gc.concept_id = ga.concept_id
+            WHERE gc.source = %s
+        );"""
+        drop_prev_symbols_query = """
+        DELETE FROM gene_previous_symbols WHERE id IN (
+            SELECT gps.id FROM gene_previous_symbols gps LEFT JOIN gene_concepts gc
+                ON gc.concept_id = gps.concept_id
+            WHERE gc.source = %s
+        );"""
+        drop_symbols_query = """
+        DELETE FROM gene_symbols WHERE id IN (
+            SELECT gs.id FROM gene_symbols gs LEFT JOIN gene_concepts gc
+                ON gc.concept_id = gs.concept_id
+            WHERE gc.source = %s
+        );"""
+        drop_xrefs_query = """
+        DELETE FROM gene_xrefs WHERE id IN (
+            SELECT gx.id FROM gene_xrefs gx LEFT JOIN gene_concepts gc
+                ON gc.concept_id = gx.concept_id
+            WHERE gc.source = %s
+        );
+        """
+        with self.conn.cursor() as cur:
+            cur.execute(drop_aliases_query, [src_name.value])
+            cur.execute(drop_associations_query, [src_name.value])
+            cur.execute(drop_prev_symbols_query, [src_name.value])
+            cur.execute(drop_symbols_query, [src_name.value])
+            cur.execute(drop_xrefs_query, [src_name.value])
+        self._drop_fkeys()
+        self._drop_indexes()
+
+        drop_concepts_query = "DELETE FROM gene_concepts WHERE source = %s;"
+        drop_source_query = "DELETE FROM gene_sources gs WHERE gs.name = %s;"
+        with self.conn.cursor() as cur:
+            cur.execute(drop_concepts_query, [src_name.value])
+            cur.execute(drop_source_query, [src_name.value])
+            self.conn.commit()
+
+        self._add_fkeys()
+        self._add_indexes()
+        self._refresh_views()
+
+    def complete_write_transaction(self) -> None:
+        """Conclude transaction or batch writing if relevant."""
+        if not self.conn.closed:
+            try:
+                self._refresh_views()
+            except UndefinedTable:
+                self.conn.rollback()
+
+    def close_connection(self) -> None:
+        """Perform any manual connection closure procedures if necessary."""
+        if not self.conn.closed:
+            self.conn.commit()
+            self.conn.close()
+
+    def load_from_remote(self, url: Optional[str]) -> None:
+        """Load DB from remote dump. Warning: Deletes all existing data. If not
+        passed as an argument, will try to grab latest release from VICC S3 bucket.
+
+        :param url: location of .tar.gz file created from output of pg_dump
+        :raise DatabaseException: if unable to retrieve file from URL or if psql
+            command fails
+        """
+        if not url:
+            url = "https://vicc-normalizers.s3.us-east-2.amazonaws.com/gene_normalization/postgresql/gene_norm_latest.sql.tar.gz"  # noqa: E501
```

Comment:
I'm not sure if we have versioning on this s3 bucket, if not we may want to consider also storing timestamped versions

Reply:
Yeah, I saved a file as "latest" since it's an easy thing to point to. If we want to get serious about this we could maybe throw together some sort of basic API that lists available dumps by date, and maybe also is in charge of storing and updating them (?) but I don't really know how that would work or if it would be worthwhile